packages/angular/src/utils/mfe/with-module-federation.ts

```diff
@@ -0,0 +1,222 @@
+  remotes?: MFERemotes;
+  exposes?: Record<string, string>;
+  shared?: (
+    libraryName: string,
+    library: SharedLibraryConfig
+  ) => SharedLibraryConfig | false;
+}
+
+function recursivelyResolveWorkspaceDependents(
+  projectGraph: ProjectGraph<any>,
+  target: string
+) {
+  let dependencies = [target];
+
+  const workspaceDependencies = (
+    projectGraph.dependencies[target] ?? []
+  ).filter((dep) => !dep.target.startsWith('npm:'));
+  if (workspaceDependencies.length > 0) {
+    for (const dep of workspaceDependencies) {
+      dependencies = [
+        ...dependencies,
+        ...recursivelyResolveWorkspaceDependents(projectGraph, dep.target),
+      ];
+    }
+  }
+  return dependencies;
+}
+
+function mapWorkspaceLibrariesToTsConfigImport(workspaceLibraries: string[]) {
+  const { projects } = new Workspaces(appRootPath).readWorkspaceConfiguration();
+  const tsConfigPath = process.env.NX_TSCONFIG_PATH ?? getRootTsConfigPath();
+  const tsConfig: ParsedCommandLine = readTsConfig(tsConfigPath);
+
+  const tsconfigPathAliases: Record<string, string[]> = tsConfig.options?.paths;
+
+  if (!tsconfigPathAliases) {
+    return workspaceLibraries;
+  }
+
+  const mappedLibraries = [];
+  for (const lib of workspaceLibraries) {
+    const sourceRoot = projects[lib].sourceRoot;
+    let found = false;
+
+    for (const [key, value] of Object.entries(tsconfigPathAliases)) {
+      if (value.find((p) => p.startsWith(sourceRoot))) {
+        mappedLibraries.push(key);
+        found = true;
+        break;
+      }
+    }
+
+    if (!found) {
+      mappedLibraries.push(lib);
+    }
+  }
+
+  return mappedLibraries;
+}
+
+async function getDependentPackagesForProject(name: string) {
+  let projectGraph: ProjectGraph<any>;
+  try {
+    projectGraph = readCachedProjectGraph();
+  } catch (e) {
+    projectGraph = await createProjectGraphAsync();
+  }
+
+  const deps = projectGraph.dependencies[name].reduce(
+    (dependencies, dependency) => {
+      const workspaceLibraries = dependencies.workspaceLibraries;
+      const npmPackages = dependencies.npmPackages;
+
+      if (dependency.target.startsWith('npm')) {
+        npmPackages.push(dependency.target.replace('npm:', ''));
+      } else {
+        workspaceLibraries.push(dependency.target);
+      }
+
+      return {
+        workspaceLibraries,
+        npmPackages,
+      };
+    },
+    { workspaceLibraries: [], npmPackages: [] }
+  );
+  deps.workspaceLibraries = deps.workspaceLibraries.reduce(
+    (workspaceLibraryDeps, workspaceLibrary) => [
+      ...workspaceLibraryDeps,
+      ...recursivelyResolveWorkspaceDependents(projectGraph, workspaceLibrary),
+    ],
+    []
+  );
+
+  deps.workspaceLibraries = mapWorkspaceLibrariesToTsConfigImport(
+    deps.workspaceLibraries
+  );
+  return deps;
+}
+
+function determineRemoteUrl(remote: string) {
+  const workspace = readWorkspaceJson();
+  return joinPathFragments(
+    workspace.projects[remote].targets.serve.options.publicHost,
```

Comment:
We set this option in the serve target in the generator. It is not foolproof, but if someone deviates then there's only so much we can do.

Reply:
That's true, but this could fail with an error like `cannot read "option" from undefined` in a scenario where that target doesn't exist which is not helpful.

I think we should document properly any constraints we have and/or add some helpful error messages when we identify something's off.

Is the usage of this function restricted to a particular executor? If so, we could search the target that's using it.

Reply:
Not necessarily. It's the internals of `withModuleFederation` which is used in webpack config.
We aren't restricting the executor. 

We could enclose it in a try-catch and provide a better error message and ensure that there are docs about the function and what it expects

Reply:
I thought as much, it shouldn't be restricted to an executor. Docs and better errors sound good to me. We can always suggest using the tuple and providing the remote url.